CondeNast/launch-vehicle-fbm
# Second pass at dispatcher / router

## Why are we doing this?

This PR adds a `dispatcher` and a `router` and moves message routing into to the latter. The idea is here is an incremental step toward having all communications between the various concerns handled via events.

There are a few motivations driving this change:

* The `src/app.js` conflates several concerns (Express routing, session management, message routing, public API for emitting events)
* Handling everything via a `dispatcher` allows third-party developers to hook into the various parts of the system by monitoring events that are triggered
* Easier testing – for example the test for the express route isn't that the various messages are routed correctly. Instead, you test for the route to emit an event and that a handler for that event is registered by default. Once that's in place you test the handler independently.
* The three preceding points can be summed up as "decoupling."


## Did you document your work?

* [x] Session documentation needs to be updated to show how to save it



## How can someone test these changes?

Steps to manually verify the change:

1. `npm install`
2. `npm test`
3. Verify that the various `router` specifications are present (current 44 passing, 2 pending)

## What possible risks or adverse effects are there?

* [x] ~~~💣  session saving is currently broken~~~
* Code churn, but that can be a good thing.
* Moving to a fully event-driven system takes a little getting used to (case in point: Mocha internals)

## What are the follow-up tasks?

* Sessions saving needs to be modified because it is happening async now
* `Messenger` needs to be extract into an object by itself
* `app.Messenger` needs to be refactored into a simple Express app that emits events everywhere
* Session code needs to be extracted into its own module / event handler

None of these are necessary for merge and are more next steps.
